### ISSUE TITLE
Is it possible and does it make sense to implement partial decompilation

### ISSUE THREAD

**Issue by KOLANICH** [leibnitz27/cfr#188]
# Problem solved by this feature
There is a library called JPype for calling Java (and JVM in general) code from python. Sometimes JVM code throws an exception. When an exception is thrown and not caught, IPython shows a stack trace with pieces of python code around problematic statements and even highlights subexpressions. It would be nice to extend this stuff to show stack traces of JVM code (decompiled to Java, of course),mif nothing better is available.

CFR is useful for that, I have figured out how to decompile some methods using the API and have even created [a wrapper](https://gitlab.com/KOLANICH/cfr.py). But I wonder if it is really needed to decompile whole methods. For example in general a method may be large, but IPython shows only about 3 lines for a stack frame. Can and should we support decompiling of only 2 lines adjacent to the problematic lines for better efficiency?

**Comment by GraxCode** [comment]
I'm sure that something can be done to only show referenced lines, but if there is no debug line number information it won't work. The problem is that there could be something on the stack at the line number that also needs to be decompiled (or just left out).

**Comment by Col-E** [comment]
> but if there is no debug line number information it won't work
Incoming beautiful debug attribute
```
LineNumberTable
  0-1: 72
  1-2: 9341
  2-3: 1
  3-4: 2147483648
  4-5: 829138
```
Now, show me lines around 2147483647 please

**Comment by GraxCode** [comment]
just go for the one with the least distance lol

**Comment by KOLANICH** [comment]
Yeah, if there is no line info, it won't work. But if there is enough of line info, is it enough to analyse only a small chunk of bytecode to recover AST around the line, or the whole method still has to be analysed anyway?

**Comment by GraxCode** [comment]
> Yeah, if there is no line info, it won't work. But if there is enough of line info, is it enough to analyse only a small chunk of bytecode to recover AST around the line, or the whole method still has to be analysed anyway?

If the lines start and end on the same depth level, it is possible. If not, then a common depth has to be used, like this:

```
line 10:
System.out.println();
while(true) {
  line 12:
  Foo.bar();
  line 13:
  ...
}
```
In this example the whole while loop has to be analyzed.

**Comment by KOLANICH** [comment]
Thanks for the info.

**Comment by leibnitz27** [comment]
[ going to leave this open for now but will close as 'no thanks' ]

I've long stated (and it's something that most of this codebase is built firmly around) that I will go to ridiculous lengths to avoid trusting metadata which is not required to be correct.   CFR is a paranoid decompiler.   Deliberately so.

The line number table is the second most extreme example of this.
(The most extreme being the local variable type table).

I'm not going to propagate line number information through the analysis - this is the sort of thing that, if present, rapidly becomes a crutch, and will allow some analyses, but significantly cripple many more.   

"What about making it optional"?  <-- nope.   If you have two major modes of operation, you will always end up testing/developing against one.  It's just nature.

There is a lengthy discussion about this in https://github.com/leibnitz27/cfr/issues/73 , and the outline I give there is really the only context in which I think it's reasonable to support line numbers.

(Provide a NEW line number table which maps bytecode to output java.  This is not expected to be retrofit to the class file, rather to be understood by downstream tools.)

**Comment by KOLANICH** [comment]
Can we map `StackTraceElement​`s to bytecode offsets other way than using the tables? If no, we can try to parse the tables ourselves.

**Comment by leibnitz27** [comment]
(BTW - I realise I'm conflating your 'partial' with 'line numbers'.  I know they're not strictly the same thing, but I think they very much map to the same problem).

It's pretty much what I say in the earlier comment - this is not a simple request (It's always the ones that *SOUND* simple that are the worst ;) ).  What needs to happen here is that all expressions need to track the bytecode that they are derived from, and a reverse line mapping needs to be generated at the end. 

What I think might be a possible way of doing this would be to expose the raw attributes in the API, in a clean way (i.e. I'm really not a huge fan of just slapping 'public' on what I have atm), and ALSO expose the 'NEW' line numbers.

BUT.... exposing the attributes would be absolutely 0 value without a way to map them back to the decompiled code.

Generating new line numbers is a big pain.  It involves lots of changes, and it's something that has been low on my queue for a long time.  I will probably do it eventually, but it's not coming imminently.
